src/ctap/timed_permissions_exp.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2019-2021 Google LLC
+//
+// Licensed under the Apache License, Version 2 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[cfg(feature = "with_ctap1")]
+use super::ClockInt;
+use embedded_time::duration::Milliseconds;
+use embedded_time::{Clock, Instant};
+
+#[derive(Debug)]
+pub enum TimedPermission<C: Clock> {
+    Waiting,
+    Granted(Instant<C>),
+}
+
+impl<C: Clock> Clone for TimedPermission<C> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Waiting => Self::Waiting,
+            Self::Granted(instant) => Self::Granted(*instant),
+        }
+    }
+}
+
+impl<C: Clock> Copy for TimedPermission<C> {}
+
+impl<C: Clock> TimedPermission<C> {
+    pub fn waiting() -> TimedPermission<C> {
+        TimedPermission::Waiting
+    }
+
+    pub fn granted(now: Instant<C>, grant_duration: Milliseconds<C::T>) -> TimedPermission<C> {
+        TimedPermission::Granted(now.checked_add(grant_duration).unwrap())
```

Comment:
embedded_time's checked_add doesn't wrap. If it would overflow, it will return None. I don't know how should we deal with it. @ia0 It is also related to your comment about wrapping. What is your opinion on dealing with it?

Reply:
I checked the [code](https://github.com/FluenTech/embedded-time/blob/83b09cb9c4e4c27fdc10e1089aab87b69a0436c9/src/instant.rs#L183) and `checked_add` actually wraps which is what we want. The error case is only when the duration is too large (which is why wrapping preserves ordering, but one shouldn't call `checked_add` multiple times though).

Reply:
Done, not granting in case of error, in new commit.